core/blockchain.go

```diff
@@ -2628,50 +2636,39 @@ func (bc *BlockChain) UpdateValidatorVotingPower(state *shard.State) error {
 		return errors.New("[UpdateValidatorVotingPower] Nil shard state")
 	}
 
-	totalEffectiveStake := make(map[uint32]numeric.Dec)
-	addrToEffectiveStakes := make(map[common.Address]map[uint32]numeric.Dec)
-
-	for _, committee := range state.Shards {
-		for _, slot := range committee.Slots {
-			if slot.TotalStake != nil {
-				if _, ok := addrToEffectiveStakes[slot.EcdsaAddress]; !ok {
-					addrToEffectiveStakes[slot.EcdsaAddress] = map[uint32]numeric.Dec{}
-				}
-				if _, ok := addrToEffectiveStakes[slot.EcdsaAddress][committee.ShardID]; !ok {
-					addrToEffectiveStakes[slot.EcdsaAddress][committee.ShardID] = numeric.NewDec(0)
-				}
-				addrToEffectiveStakes[slot.EcdsaAddress][committee.ShardID] = addrToEffectiveStakes[slot.EcdsaAddress][committee.ShardID].Add(*slot.TotalStake)
+	rosters := make([]votepower.RosterPerShard, len(state.Shards))
 
-				if _, ok := totalEffectiveStake[committee.ShardID]; !ok {
-					totalEffectiveStake[committee.ShardID] = numeric.NewDec(0)
-				}
-				totalEffectiveStake[committee.ShardID] = totalEffectiveStake[committee.ShardID].Add(*slot.TotalStake)
-			}
+	for i := range state.Shards {
+		roster, err := votepower.Compute(state.Shards[i].Slots)
+		if err != nil {
+			return err
 		}
+		rosters[i] = votepower.RosterPerShard{state.Shards[i].ShardID, roster}
 	}
 
-	batch := bc.db.NewBatch()
-	for addr, votingPowers := range addrToEffectiveStakes {
-		addrTotalVotingPower := numeric.NewDec(0) // Total voting power is the average voting power across all shards
-		addrTotalEffectiveStake := numeric.NewDec(0)
-		for shardID, eStake := range votingPowers {
-			addrTotalVotingPower = addrTotalVotingPower.Add(eStake.Quo(totalEffectiveStake[shardID]))
-			addrTotalEffectiveStake = addrTotalEffectiveStake.Add(eStake)
-		}
-		// Retrieve the stats and update
-		stats, err := rawdb.ReadValidatorStats(bc.db, addr)
-		if stats == nil {
-			stats = &staking.ValidatorStats{big.NewInt(0), big.NewInt(0), big.NewInt(0), numeric.NewDec(0), numeric.NewDec(0)}
-		}
+	networkWide := votepower.AggregateRosters(rosters)
 
-		stats.AvgVotingPower = addrTotalVotingPower.Quo(numeric.NewDec(int64(len(state.Shards))))
-		stats.TotalEffectiveStake = addrTotalEffectiveStake
+	batch := bc.db.NewBatch()
 
-		err = rawdb.WriteValidatorStats(batch, addr, stats)
+	for key, value := range networkWide {
+		statsFromDB, err := rawdb.ReadValidatorStats(bc.db, key)
+		if statsFromDB == nil {
+			statsFromDB = &staking.ValidatorStats{
+				big.NewInt(0), big.NewInt(0), big.NewInt(0),
+				numeric.NewDec(0), numeric.NewDec(0),
+				[]staking.VotePerShard{}, []staking.KeysPerShard{},
+			}
+		}
+		statsFromDB.AvgVotingPower = value.AverageVotingPower
```

Comment:
This can be removed. You can counting the avg voting power per slot, which is not interesting. It should be the avg of powers in four different shards, which can be computed in the VotingPowerPerShard. So let's just remove this field.